### PR TITLE
Add raw http test with response with adding extra header

### PR DESCRIPTION
**Pull request by mkudlej**
closes https://github.com/Kuadrant/testsuite/issues/9

### CODE REVIEW

File: testsuite/tests/kuadrant/authorino/operator/http/conftest.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+from testsuite.httpx import HttpxBackoffClient
+
+HTTP_DEFAULT_PORT=5001
+
+# pylint: disable=unused-argument
+@pytest.fixture(scope="module")
+def authorization(authorization, authorino, wildcard_domain, blame, openshift, label, rhsso_service_info) -> Authorization:
+    """In case of Authorino, AuthConfig used for authorization"""
+    #if authorization is None:
+    authorization = AuthConfig.create_instance(openshift, blame("ac"), wildcard_domain, labels={"testRun": label})
```
Comment (pehala) [inline testsuite/tests/kuadrant/authorino/operator/http/conftest.py:17]: Use module_label instead, the label is for session scoped.

File: testsuite/tests/kuadrant/authorino/operator/http/conftest.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
+"""Conftest for all tests requiring custom deployment of Authorino"""
+import pytest
+from weakget import weakget
+
+from testsuite.openshift.objects.authorino import AuthorinoCR
+from testsuite.objects import Authorino, Authorization
+from testsuite.openshift.objects.auth_config import AuthConfig
+from testsuite.httpx import HttpxBackoffClient
+
+HTTP_DEFAULT_PORT=5001
```
Comment (pehala) [inline testsuite/tests/kuadrant/authorino/operator/http/conftest.py:10]: I believe the services have their ports named so you could expose the route and use a name instead an actual port

File: testsuite/tests/kuadrant/authorino/operator/http/conftest.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    yield client
+    client.close()
+    authorization.delete()
+
+
+@pytest.fixture(scope="module")
+def authorino_route(authorino, blame, openshift):
+    name = f"route-{authorino.name()}"
+    # expose cannot be used, see https://github.com/openshift/origin/issues/8699
+    route = openshift.do_action('create', 'route', 'edge', name,
```
Comment (pehala) [inline testsuite/tests/kuadrant/authorino/operator/http/conftest.py:40]: I do not understand why expose won't work, because that issue seems to be closed. I believe binding it to a specific service port should work

Comment (pehala) [reply testsuite/tests/kuadrant/authorino/operator/http/conftest.py:40]: Also, you could try using `openshift.routes.expose`

File: testsuite/tests/kuadrant/authorino/operator/http/test_raw_http.py [hunk 4]
```diff
@@ -0,0 +1,13 @@
+"""
+Test raw http authorization interface.
+"""
+
+import pytest
+
+def test_first(authorization, client_http_auth, auth):
+    response = client_http_auth.request("GET", "/check", auth=auth, json='{}')
+    # ako sa spozna, ze komunikoval s rhsso a overil to?
```
Comment (pehala) [inline testsuite/tests/kuadrant/authorino/operator/http/test_raw_http.py:9]: You can add tests with invalid token and without token at all

File: testsuite/tests/kuadrant/authorino/operator/http/test_raw_http.py [hunk 5]
```diff
@@ -0,0 +1,13 @@
+"""
+Test raw http authorization interface.
+"""
+
+import pytest
+
+def test_first(authorization, client_http_auth, auth):
```
Comment (pehala) [inline testsuite/tests/kuadrant/authorino/operator/http/test_raw_http.py:7]: Please change testname to something more explicit

File: testsuite/tests/kuadrant/authorino/operator/tls/conftest.py [hunk 6]
```diff
@@ -1,5 +1,4 @@
 """Conftest for all TLS-enabled tests"""
-
```
Comment (pehala) [inline testsuite/tests/kuadrant/authorino/operator/tls/conftest.py:2]: Unnecessary change

File: testsuite/tests/kuadrant/authorino/operator/http/conftest.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+HTTP_DEFAULT_PORT=5001
+
+# pylint: disable=unused-argument
+@pytest.fixture(scope="module")
+def authorization(authorization, authorino, wildcard_domain, blame, openshift, label, rhsso_service_info) -> Authorization:
+    """In case of Authorino, AuthConfig used for authorization"""
+    #if authorization is None:
+    authorization = AuthConfig.create_instance(openshift, blame("ac"), wildcard_domain, labels={"testRun": label})
+    authorization.add_oidc_identity("rhsso", rhsso_service_info.issuer_url())
+    authorization.add_response(inject_json={'name': 'another-json-returned-in-a-header',
```
Comment (pehala) [inline testsuite/tests/kuadrant/authorino/operator/http/conftest.py:19]: I might be missing something, but `add_response` doesn't seem to have a parameter called `inject_json`

File: testsuite/tests/kuadrant/authorino/operator/http/conftest.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+                                'json': { 'properties': [{'name': 'propX', 'value': 'valueX'}]}
+                                })
+    return authorization
+
+
+@pytest.fixture(scope="module")
+def client_http_auth(authorization, authorino_route):
+    """Returns httpx client to be used for requests, it also commits AuthConfig"""
+    authorization.commit()
+    client = HttpxBackoffClient(base_url=f"https://{authorino_route.as_dict()['spec']['host']}", verify=False)
```
Comment (pehala) [inline testsuite/tests/kuadrant/authorino/operator/http/conftest.py:30]: `authorino_route.model.spec.host`

**Comment by pehala** [comment]
The reason, why the acceptance checks fail, is that you have added `__init__.py` to the root directory, where it shouldn't be and you are missing it in your `tests/kuadrant/authorino/operator/http` dir, where it should be

File: testsuite/tests/kuadrant/authorino/operator/http/conftest.py (hunk 3 above)
Comment (mkudlej) [reply testsuite/tests/kuadrant/authorino/operator/http/conftest.py:40]: It still doesn't work, see:
`$ oc expose service --name=route-authorin-mkudlej-mkdlej-kd-bzb23 -o json authorin-mkudlej-mkdlej-kd-bzb2-authorino-authorization --target-port=5001 --port=80

error: cannot use --target-port with --generator=route/v1`
I don't know what do you mean by `Also, you could try using openshift.routes.expose`, there is no such submodule of openshift client named `routes.expose` or maybe I haven't found it.

File: testsuite/tests/kuadrant/authorino/operator/http/test_raw_http.py [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+
+def test_authorized_via_http(authorization, client_http_auth, auth):
+    response = client_http_auth.request("GET", "/check", auth=auth, json='{}')
+    assert response.is_success
+    assert response.status_code == 200
+    assert response.text == ''
+    assert response.headers.get('x-ext-auth-other-json', '') == '{"propX":"valueX"}'
+
+def test_unauthorized_via_http(authorization, client_http_auth):
+    response = client_http_auth.request("GET", "/check", json='{}')
```
Comment (pehala) [inline testsuite/tests/kuadrant/authorino/operator/http/test_raw_http.py:15]: `client_http_auth.get`

File: testsuite/tests/kuadrant/authorino/operator/http/test_raw_http.py [hunk 10]
```diff
@@ -0,0 +1,18 @@
+"""
+Test raw http authorization interface.
+"""
+
+import pytest
+
+def test_authorized_via_http(authorization, client_http_auth, auth):
+    response = client_http_auth.request("GET", "/check", auth=auth, json='{}')
+    assert response.is_success
```
Comment (pehala) [inline testsuite/tests/kuadrant/authorino/operator/http/test_raw_http.py:9]: Redundant since you are checking the status code anyway

File: testsuite/tests/kuadrant/authorino/operator/http/test_raw_http.py [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+    assert response.status_code == 200
+    assert response.text == ''
+    assert response.headers.get('x-ext-auth-other-json', '') == '{"propX":"valueX"}'
+
+def test_unauthorized_via_http(authorization, client_http_auth):
+    response = client_http_auth.request("GET", "/check", json='{}')
+    assert not response.is_success
```
Comment (pehala) [inline testsuite/tests/kuadrant/authorino/operator/http/test_raw_http.py:16]: Redundant

File: testsuite/tests/kuadrant/authorino/operator/http/test_raw_http.py [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+    assert response.status_code == 401
+    assert response.text == ''
```
Comment (pehala) [inline testsuite/tests/kuadrant/authorino/operator/http/test_raw_http.py:18]: The text is important in this one, right?

File: testsuite/tests/kuadrant/authorino/operator/http/conftest.py [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+from testsuite.openshift.objects.authorino import AuthorinoCR
+from testsuite.objects import Authorino, Authorization
+from testsuite.openshift.objects.auth_config import AuthConfig
+from testsuite.httpx import HttpxBackoffClient
+
+# pylint: disable=unused-argument
+@pytest.fixture(scope="module")
+def authorization(wildcard_domain, blame, openshift, rhsso_service_info, module_label) -> Authorization:
+    """In case of Authorino, AuthConfig used for authorization"""
+    authorization = AuthConfig.create_instance(openshift, blame("ac"), wildcard_domain, labels={"testRun": module_label})
```
Comment (pehala) [inline testsuite/tests/kuadrant/authorino/operator/http/conftest.py:14]: There is a different solution that doesn´t require creating new `AuthConfig` and using existing one. (I came up with that only after we talked, sorry)

```
authorization.remove_all_hosts()
authorization.add_host(wildcard_domain)
```

File: testsuite/tests/kuadrant/authorino/operator/http/test_raw_http.py (hunk 12 above)
Comment (mkudlej) [reply testsuite/tests/kuadrant/authorino/operator/http/test_raw_http.py:18]: What do you mean exactly? I think it is needed to check as much as we can. On the other hand Authorino returns some explaining text in case of not successful call. Do you think that text is missing in this case?

Comment (pehala) [reply testsuite/tests/kuadrant/authorino/operator/http/test_raw_http.py:18]: Sorry, the question should be `The text is NOT important in this one, right?`. Authorino uses Header for extra information about checks not response text I believe, so this check is not needed.

File: testsuite/tests/kuadrant/authorino/operator/http/test_raw_http.py (hunk 9 above)
Comment (mkudlej) [reply testsuite/tests/kuadrant/authorino/operator/http/test_raw_http.py:15]: `TypeError: Client.get() got an unexpected keyword argument 'json'`

Comment (pehala) [reply testsuite/tests/kuadrant/authorino/operator/http/test_raw_http.py:15]: Why are you sending empty JSON in a get request?

Comment (mkudlej) [reply testsuite/tests/kuadrant/authorino/operator/http/test_raw_http.py:17]: See this example:
```kubectl port-forward services/authorino-authorino-authorization 3000:3000 &;PID=$!

curl -H 'Content-type: application/json' -d '{}' http://wristband.127.0.0.1.nip.io:3000/check -i
# HTTP/1.1 200 OK

kill $PID``` from https://github.com/Kuadrant/authorino-operator/pull/61

File: testsuite/tests/kuadrant/authorino/operator/http/conftest.py [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+    """Returns httpx client to be used for requests, it also commits AuthConfig"""
+    client = HttpxBackoffClient(base_url=f"http://{authorino_route.model.spec.host}", verify=False)
+    yield client
+    client.close()
+
+
+@pytest.fixture(scope="module")
+def authorino_route(authorino, blame, openshift):
+    """Add route for authorino http port to be able to access it."""
+    name = f"route-{authorino.name()}"
```
Comment (pehala) [inline testsuite/tests/kuadrant/authorino/operator/http/conftest.py:34]: Use blame instead, this could fail if multiple people are running this test

File: testsuite/tests/kuadrant/authorino/operator/http/conftest.py [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+            'wrapperKey': 'x-ext-auth-other-json',
+            'json': {'properties': [
+                {'name': 'propX', 'value': 'valueX'}
+            ]}}
+    authorization.add_response(response=resp)
+    return authorization
+
+
+@pytest.fixture(scope="module")
+def client_http_auth(authorization, authorino_route):
```
Comment (pehala) [inline testsuite/tests/kuadrant/authorino/operator/http/conftest.py:24]: Can you rename this to simply `client`?

Comment (pehala) [reply testsuite/tests/kuadrant/authorino/operator/http/conftest.py:24]: And that's exactly why I want this to be renamed. It acts exactly as a client and provides the same functionality. The original fixture doesnt work in this context as there is no envoy here, so we can safely replace it by this one which does work and is used for the same purpose.

Comment (mkudlej) [reply testsuite/tests/kuadrant/authorino/operator/http/conftest.py:24]: ok, if you think this is a good idea, then I rename it.

**Review by pehala — APPROVED** [review]

**Review by JaurbanRH — APPROVED** [review]
